Edit task

LucasEduardoGiovanini opened the pull request:
About the issue, i implemented  methods to:
 - update a task
 - update the list name 
 - order tasks by priority level.
About the tasks, i created a method in the class TodoList that allows you to create a task and add it in the respective list.

LucasEduardoGiovanini committed:
apply pep8 standard and create edit function

if run, this commit will refactor some functions name to enter the pep8 standard and create a function edit in the task package.

LucasEduardoGiovanini committed:
refactor some functions name to apply the pep8 pattern

LucasEduardoGiovanini committed:
creatiing edit list functions

if run, this commit will create functions to edit tasks and also to edit the list name.

LucasEduardoGiovanini committed:
Creating functions undo and complete task in package todolist

if run, this commit will create functions to make it possible to undo or complete a task by the TodoList class.

LucasEduardoGiovanini committed:
refactoring some functions of the tests package

if run, this commit will refactor the test functions to access a task into a list, interrupting operations directly on the task.

LucasEduardoGiovanini committed:
implements tests about update a task.

LucasEduardoGiovanini committed:
implement a test about update the list name.

LucasEduardoGiovanini committed:
include pep8 pattern

LucasEduardoGiovanini committed:
creating a function to order tasks by priority and create task by list

if run, this commit will create a function to order task by all priorities and make my TodoList class create the tasks.

LucasEduardoGiovanini committed:
[1/2] refactoring the tests.

if run, this commit will refactor the tests to generate the tasks using the TodoList Method create_task.

LucasEduardoGiovanini committed:
[2/2] refactoring the tests.

LucasEduardoGiovanini committed:
creating the test function  to order tasks by priority value

LucasEduardoGiovanini commented:
About delete the todo list, i can create a test about and call `del mylist` ? because i can't delete it in your own class.

LucasEduardoGiovanini committed:
Update README

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -17,6 +17,17 @@ def remove_task(self, task:Task):
+    def edit_task(self,task:Task, name=None, description=None, completed=None, priority=None):
+        try:
+            self.tasks[self.tasks.index(task)].edit(name,description,completed,priority)
+        except ValueError as ve:
```
In addition to `ValueError`, this function can also raise `IndexError`.

johnnywell replied:
I was not talking about te name you are using for the Exception, but the exception this function could raise and you are not handling it.

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -9,6 +9,23 @@ def add_task(self, task:Task):
+    def undo_task(self,task:Task):
+        try:
+            self.tasks[self.tasks.index(task)].undo()
+        except ValueError as ve:
```
In addition to `ValueError`, this function can also raise` IndexError`.

johnnywell replied:
I was not talking about te name you are using for the Exception, but the exception this function could raise and you are not handling it.

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -9,6 +9,23 @@ def add_task(self, task:Task):
+    def complete_task(self,task:Task):
+        try:
+            self.tasks[self.tasks.index(task)].complete()
+        except ValueError as ve:
```
In addition to `ValueError`, this function can also raise` IndexError`.

johnnywell replied:
I was not talking about te name you are using for the Exception, but the exception this function could raise and you are not handling it.

LucasEduardoGiovanini replied:
Yes, I just renamed the class because I thought of creating another one for the index error, but I ended up realizing that it was not necessary.

LucasEduardoGiovanini replied:
sorry, I should have explained my intention at the commit.

LucasEduardoGiovanini replied:
But soon I will create an error class to deal with improper types of variables, for example:  passing as a parameter a `string` for `priority`attribute, which needs an `integer`parameter.

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -50,6 +55,10 @@ def edit_list_name(self, name=None):
     def __str__(self):
         return "nome lista: " + self.name + " tarefas: " + str(self.tasks)
 
+    def order_task_priority(self):
```
I think the name you're looking for is `sort_tasks_by_priority`.

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -50,6 +55,10 @@ def edit_list_name(self, name=None):
         return "nome lista: " + self.name + " tarefas: " + str(self.tasks)
 
+    def order_task_priority(self):
+        self.tasks = sorted(self.tasks, key=operator.attrgetter("priority"))
```
kudos for the use of sorted and attrgetter. 👍

johnnywell commented:
> About delete the todo list, i can create a test about and call `del mylist` ? because i can't delete it in your own class.

Yes you can, just make sure the tasks are deleted too.

LucasEduardoGiovanini committed:
renaming the method name to make it coherent

LucasEduardoGiovanini committed:
renaming the method name to make it coherent

LucasEduardoGiovanini committed:
refactoring some todo list methods for catch the Indexerror exception and renaming the custom exception to its old name.

LucasEduardoGiovanini committed:
Creating a id attribute for Tasks and use case and create getters and setters.

if run, this commit will create the attribute commented above and create getters and setters for the attributes: completed and priority.
this getters and setters allow attributes to be set with just one type of parameter

LucasEduardoGiovanini committed:
creating a treatment for improper type of parameter and testing it

if run, this commit will create a if stantment to check that the type of parameters are correct. If not, i create a message error and call the custom error class.

The getters and setters are working only if the user call the print of todolist, but i want to check before the object creation if all parameters is ok.

LucasEduardoGiovanini committed:
refactor the treatment for improper type of parameter and making the test about priority parameter.

If run, this commit will refactor the if statement to permit all types for name and description of a task. Also i create a test to insert a not correct parameter type to priority attribute.

LucasEduardoGiovanini committed:
refactoring the error message.

LucasEduardoGiovanini commented:
About delete a todo list, i will leave it to do when i create my use case package. I tried to do that in the test package, but it wasn't getting nice.

LucasEduardoGiovanini committed:
making the getters and setters treat invalid type of parameters.

if run, this commit will remove the if statement and delegate the verification of parameter type for setters.
Before, the setters couldn't validate because i was calling the custom exception in wrong place.

LucasEduardoGiovanini commented:
At the time, none of my methods use the ID attribute  to perform some operation. Because the application don't use a database in this moment, so i don't need to recover nothing, all objects remain created while the application is running.
Taking this into account, can I continue performing my methods using the created object? Or should i adapt the methods to receive the object ID?
For example:
**how is it now**
`task1.edit(name="johnny")`

**how could it stay**
```
list_of_tasks(task1, task2, task3)

Task.edit(id_task, name="johnny) #edit search the task in the list_of_tasks using the given ID.

```

johnnywell commented on domain/entities/task.py:
```diff
@@ -1,17 +1,59 @@
+    def completed(self):
+        try:
+            return self._completed
+        except Exception as e:
```
What can make this exception to happen? I Think the message you are passing is not helpful to understand the problem.

johnnywell commented on domain/entities/task.py:
```diff
@@ -1,17 +1,59 @@
+    def priority(self):
+        try:
+            return self._priority
+        except Exception as e:
```
What can make this exception to happen? I Think the message you are passing is not helpful to understand the problem.

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -1,28 +1,76 @@
         except ValueError as ve:
-            raise todoListError(ve, 13, "removing a task that does not exist in this list")
+            raise TodoListError(ve, 38, "removing a task that does not exist in this list")
+        except IndexError as ie:
```
I think this exception won't happen.

LucasEduardoGiovanini replied:
I was thinking if the application is running on multiple machines, more than one user can try to delete a task. Doesn't it make a lot of sense to do so?

johnnywell replied:
When running in multiple machines the point of consistency is the database, not memory. But concurrent requests on the same machine could make some sense if they are trying tho access memory from other threads. In any case I don't think a IndexError will ever be raised when trying to remove an element from a list using the `remove` method.

LucasEduardoGiovanini replied:
got it, thanks for the help!

LucasEduardoGiovanini committed:
applying the requested changes.
